backend/src/apiserver/resource/resource_manager.go

```diff
@@ -229,6 +236,117 @@ func (r *ResourceManager) GetPipelineTemplate(pipelineId string) ([]byte, error)
 	return template, nil
 }
 
+// This only detects default usage of use_gcp_secret sdk method.
+func isGcpSecretVolume(volume apicorev1.Volume) bool {
+	return volume.Secret != nil &&
+		volume.Name == defaultGcpSecretVolumeName &&
+		volume.Secret.SecretName == defaultGcpSecretName &&
+		(volume.Secret.Optional == nil || !*volume.Secret.Optional)
+}
+
+func isGcpSecretVolumeMount(volumeMount apicorev1.VolumeMount) bool {
+	return volumeMount.Name == defaultGcpSecretVolumeName
+}
+
+// user-gcp-sa secret was the default way to authenticate in GCP, but it is no longer needed when
+// we suggest using default authentication methods provided by GKE.
+// Alternatives include: project default service account, node pool default service account and
+// workload identity.
+// Dropping GOOGLE_APPLICATION_CREDENTIALS env when the secret is not configured ensures old
+// pipelines can still be run in new recommended clusters set up as above methods.
+// Reference, old pipeline mounts secret like https://github.com/kubeflow/pipelines/blob/e19efdff76829ee374f400e68cc478ebff53a845/sdk/python/kfp/gcp.py#L46
+//
+// TODO: when all pipeline samples no longer use user-gcp-sa secret, we can remove this usage.
+func (r *ResourceManager) dropUserGcpSaIfNotConfigured(workflow util.Workflow) error {
+	someStepHasGcpSecretVolume := false
+	for _, template := range workflow.Spec.Templates {
+		for _, volume := range template.Volumes {
+			if isGcpSecretVolume(volume) {
+				someStepHasGcpSecretVolume = true
+				break
+			}
+		}
```

Comment:
nit: here also can break
if someStepHasGcpSecretVolume is true, it also can break in outer "for".

Reply:
In fact, I would prefer writing no breaks here to keep it simple and easy to read. (because this cannot impact performance)

WDYT?

Reply:
sure